src/fb.c

```diff
@@ -40,24 +43,147 @@ static inline uint32_t rgbx_to_rgb30(u32 c)
+}
+
+static u32 font_get_pixel(u8 c, u32 x, u32 y)
+{
+    c -= 0x20;
+    u8 *ptr = &_binary_font_b_start[c * 512 + (y * 8 + x) * 4];
+    u8 r = *ptr++;
+    u8 g = *ptr++;
+    u8 b = *ptr++;
+
+    return (b << 2) | (g << 12) | (r << 22);
+}
+
+static void fb_console_blit_char(console_pos_t pos)
+{
+    u32 x = 30 + pos.col * 8;
+    u32 y = 50 + pos.row * 16;
+    u8 c = console.text[fb_console_get_row(pos.row)][pos.col];
+
+    if (!console.initialized)
+        return;
+
+    if (c == '\0') {
+        fb_fill(x, y, 8, 16, 0);
+    } else {
+        for (int i = 0; i < 16; i++)
+            for (int j = 0; j < 8; j++)
+                fb_set_pixel(x + j, y + i, font_get_pixel(c, j, i));
+    }
+}
+
+static void fb_console_blit_row(u32 row)
+{
+    console_pos_t pos = {.row = row, .col = 0};
+
+    while (pos.col < CONSOLE_MAX_COLS) {
+        fb_console_blit_char(pos);
+        pos.col++;
+    }
+}
+
+static void fb_console_blit_all(void)
+{
+    for (u32 row = 0; row < CONSOLE_MAX_ROWS; ++row)
+        fb_console_blit_row(row);
+}
+
+static void fb_console_putbyte(u8 c)
+{
+    if (console.cursor.col >= CONSOLE_MAX_COLS)
+        return;
+
+    console.text[fb_console_get_row(console.cursor.row)][console.cursor.col] = c;
+    fb_console_blit_char(console.cursor);
+
+    console.cursor.col++;
+}
+
+static void fb_check_scroll(void)
+{
+    if (console.cursor.row < CONSOLE_MAX_ROWS)
+        return;
+
+    memset(console.text[fb_console_get_row(0)], 0, CONSOLE_MAX_COLS);
+    console.row_offset = (console.row_offset + 1) % CONSOLE_MAX_ROWS;
+    console.cursor.row--;
+
+    assert(console.cursor.row < CONSOLE_MAX_ROWS);
+
+    fb_console_blit_all();
+}
+
+void fb_console_putc(u8 c)
+{
+    if (c == '\n') {
+        console.cursor.row++;
+        console.cursor.col = 0;
+    } else if (c == '\r') {
+        console.cursor.col = 0;
+    } else if (c >= 0x20 && c <= 0x7e) {
+        fb_console_putbyte(c);
+    }
+
+    fb_check_scroll();
+}
+
+void fb_console_write(const void *buf, size_t count)
+{
+    const u8 *p = buf;
+
+    while (count--)
+        fb_console_putc(*p++);
+}
+
 void fb_blit(int x, int y, int w, int h, void *data, int stride)
 {
     uint32_t *p = data;
 
     for (int i = 0; i < h; i++)
         for (int j = 0; j < w; j++)
-            fb[x + j + (y + i) * fb_s] = rgbx_to_rgb30(p[j + i * stride]);
+            fb_set_pixel(x + j, y + i, rgbx_to_rgb30(p[j + i * stride]));
 }
 
 void fb_fill(int x, int y, int w, int h, u32 color)
 {
     for (int i = 0; i < h; i++)
         for (int j = 0; j < w; j++)
-            fb[x + j + (y + i) * fb_s] = rgbx_to_rgb30(color);
+            fb_set_pixel(x + j, y + i, rgbx_to_rgb30(color));
 }
 
 void fb_display_logo(void)
 {
     printf("fb: display logo\n");
-    fb_blit((fb_w - logo_w) / 2, (fb_h - logo_h) / 2, logo_w, logo_h, logo, logo_w);
+    fb_blit((fb.w - logo.w) / 2, (fb.h - logo.h) / 2, logo.w, logo.h, logo.ptr, logo.w);
 }
+
+void fb_init(void)
```

Comment:
Because I eventually need to call `fb_console_blit_all();` after the frame buffer is initialised to show the text that was printed before. I could either add a prototype for `fb_console_blit_all()` at the beginning or just move the init further below. I'll fix this in the header for consistency as well.